### PR TITLE
asserts: support parsing the plugs stanza i.e. plug rules in snap-declarations

### PR DESCRIPTION
**Pull request by pedronis**

### CODE REVIEW

**Comment by pedronis** [comment]
based on #2015

File: asserts/ifacedecls.go [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+	case "slot-attributes":
+		c.SlotAttributes = cstrs
+	default:
+		panic("unknown PlugConnectionConstraints field " + field)
+	}
+}
+
+func (c *PlugConnectionConstraints) setIDConstraints(field string, cstrs []string) {
+	switch field {
+	case "slot-snap-id":
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Can we please have slot-snap-type here and plug-snap-type on the other end? Also lists.

File: asserts/ifacedecls.go [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+	if cMap == nil {
+		if shortcut {
+			return &PlugConnectionConstraints{PlugAttributes: AlwaysMatchAttributes, SlotAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugConnectionConstraints{PlugAttributes: NeverMatchAttributes, SlotAttributes: NeverMatchAttributes}, nil
+	}
+	plugConnCstrs := &PlugConnectionConstraints{}
+	defaultUsed := 0
+	for _, field := range plugIDConstraints {
+		// TODO: check these further against regexps?
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Yeah, might be a good idea to ensure we can use special keys in the future ($same or the like) without being afraid we're blowing something up.

File: asserts/ifacedecls.go [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+			if err != nil {
+				return nil, fmt.Errorf("cannot compile %s in %s: %v", field, context, err)
+			}
+		} else {
+			defaultUsed++
+		}
+		plugConnCstrs.setAttributeConstraints(field, cstrs)
+	}
+	if defaultUsed == len(attributeConstraints)+len(plugIDConstraints) {
+		return nil, fmt.Errorf("%s must specify at least one of %s, %s", context, strings.Join(attributeConstraints, ", "), strings.Join(plugIDConstraints, ", "))
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Nice.

File: asserts/ifacedecls.go [hunk 4]
```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+	result error
+}
+
+func (matcher fixedAttrMatcher) match(context string, v interface{}) error {
+	return matcher.result
+}
+
+var (
+	AlwaysMatchAttributes = &AttributeConstraints{matcher: fixedAttrMatcher{nil}}
+	NeverMatchAttributes  = &AttributeConstraints{matcher: fixedAttrMatcher{errors.New("not allowed, never matching")}}
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Will that ", never matching" be surfacing to users?  People might be confused as they don't know the decision process goes through such a mechanism. Perhaps just "not allowed" would be better if that's the case.

File: asserts/ifacedecls.go [hunk 5]
```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+		switch x {
+		case "true":
+			return nil, true, nil
+		case "false":
+			return nil, false, nil
+		default:
+			return nil, false, fmt.Errorf("%s is not a map but is set to a shortcut that is not 'true' nor 'false': %q", context, x)
+		}
+	default:
+		return nil, false, fmt.Errorf("%s must be a map or one of the shortcuts 'true' or 'false'", context)
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: This message seems nicer than the previous one, for both cases. Perhaps we can just remove both defaults and move that return to the end of the function.

File: asserts/ifacedecls.go [hunk 6]
```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+	if err != nil {
+		return nil, err
+	}
+	if cMap == nil {
+		if shortcut {
+			return &PlugInstallationConstraints{PlugAttributes: AlwaysMatchAttributes}, nil
+		}
+		return &PlugInstallationConstraints{PlugAttributes: NeverMatchAttributes}, nil
+	}
+	attrs, err := compileAttributeConstraints(cMap["plug-attributes"])
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Can we please also introduce plug-snap-type and slot-snap-type on the opposite end? Also lists.

File: asserts/ifacedecls.go [hunk 7]
```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+)
+
+func compilePlugRule(interfaceName string, rule interface{}) (*PlugRule, error) {
+	context := fmt.Sprintf("plug rule for interface %q", interfaceName)
+	rMap, shortcut, err := checkMapOrShortcut(context, rule)
+	if err != nil {
+		return nil, err
+	}
+	if rMap == nil {
+		if shortcut {
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Hmmm, sounds reasonable I guess. We need to tweak the docs.

File: asserts/ifacedecls_test.go [hunk 8]
```diff
@@ -268,3 +273,239 @@ foo: [{p: "zzz"}, {p: "/foo/y"}]
...
+	checkBoolPlugConnConstraints(c, rule.AllowAutoConnection, true)
+	// ... but deny-auto-connection is on
+	checkBoolPlugConnConstraints(c, rule.DenyAutoConnection, true)
+}
+
+func (s *plugSlotRulesSuite) TestCompilePlugRuleConnectionConstraintsIDConstraints(c *C) {
+	rule, err := asserts.CompilePlugRule("iface", map[string]interface{}{
+		"allow-connection": map[string]interface{}{
+			"slot-snap-id":      []interface{}{"snapid1", "snapid2"},
+			"slot-publisher-id": []interface{}{"canonical", "$same"},
```
Comment (niemeyer) [inline asserts/ifacedecls_test.go]: Instead of $same, it might be more useful to have something like $plug_account_id and $slot_account_id, so it's both clear and we can reuse the concept in other fields if necessary, without having to use different language on each.

It won't make a difference here, though. Just brainstorming.

File: asserts/ifacedecls.go [hunk 9]
```diff
@@ -220,7 +222,256 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+	}
+	plugRule := &PlugRule{
+		Interface: interfaceName,
+	}
+	defaultUsed := 0
+	// installation subrules
+	for _, subrule := range installationSubrules {
+		v := rMap[subrule]
+		if v == nil {
+			v = defaultAllowRule[subrule]
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: I had some difficulty reading this part of the logic, and I think the problem is the name of this variable. This mixes the idea of "default" and "allow", when some of those rules are themselves named allow, and others are deny. so the "default allow of deny".. (brain explodes)

Perhaps "defaultOutcome" and "invertedOutcome"?

File: asserts/ifacedecls.go (hunk 7 above)
Comment (niemeyer) [inline asserts/ifacedecls.go]: But maybe I misunderstand. If map is nil, won't we necessarily hit the defalutUsed error at the end? How can this be useful if so?

File: asserts/ifacedecls.go (hunk 4 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: I'm not sure myself atm, may be safer indeed to just make it "now allowed" for now

File: asserts/ifacedecls.go (hunk 5 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: ok

File: asserts/ifacedecls.go (hunk 6 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: ok

File: asserts/ifacedecls.go (hunk 9 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: those names work for me

File: asserts/ifacedecls.go (hunk 1 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: will do

File: asserts/ifacedecls.go (hunk 7 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: this makes iface: true|false => correspond to iface: defaultOutcome|invertedOutcome

and indeed is a generalisation not mentioned in the doc, but it seems useful for the base-declaration if I didn't misunderstand the idea that we want to list all interfaces

Comment (niemeyer) [inline asserts/ifacedecls.go]: Yeah, that's what I understood at first. My second comment above is the one that is off. On re-review I see it won't fall onto the defaultUsed case.

File: asserts/ifacedecls.go [hunk 10]
```diff
@@ -434,10 +441,10 @@ func compilePlugRule(interfaceName string, rule interface{}) (*PlugRule, error)
 		return nil, err
 	}
 	if rMap == nil {
-		if shortcut {
-			rMap = defaultAllowRule
-		} else {
-			rMap = denyRule
+		if shortcut == "true" {
+			rMap = defaultOutcome
+		} else { // "false"
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Why the change out of bool?  It looks like we're using it as a bool still, even assuming the alternative value when it's not "true", so it's not clear what we won here?

Comment (pedronis) [inline asserts/ifacedecls.go]: I had the impression that with shortcut bool, it wasn't clear whether it was carrying the shortcut value or telling there was shortcut

Comment (niemeyer) [inline asserts/ifacedecls.go]: Perhaps "invert bool" is a better variable here, and makes the association with invertOutcome more clear. It's also nice that it's false by default, which means it only changes something if set.

Comment (pedronis) [inline asserts/ifacedecls.go]: as discussed online will try this change in the follow up #2035

File: asserts/ifacedecls.go [hunk 11]
```diff
@@ -143,7 +145,7 @@ type regexpAttrMatcher struct {
 }
 
 func compileRegexpAttrMatcher(cc compileContext, s string) (attrMatcher, error) {
-	rx, err := regexp.Compile("^" + s + "$")
+	rx, err := regexp.Compile("^(" + s + ")$")
```
Comment (jdstrand) [inline asserts/ifacedecls.go]: This is a nice catch. It means that now 'foobar|z' correctly matches 'foobar' and 'z' and that you need to use 'fooba(r|z)' to match 'foobar' and 'foobaz'.

File: asserts/ifacedecls.go [hunk 12]
```diff
@@ -220,7 +222,263 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+			return nil, "true", nil
+		case "false":
+			return nil, "false", nil
+		}
+	}
+	return nil, "", fmt.Errorf("%s must be a map or one of the shortcuts 'true' or 'false'", context)
+
+}
+
+// PlugRule holds the rule of what is allowed, wrt installation and connection, for a plug of a specific interface for a snap.
```
Comment (jdstrand) [inline asserts/ifacedecls.go]: Minor nitpick-- I prefer when comments are wrapped to 80 characters (really, I prefer when everything is, but I realize that isn't the style here). This is especially true of github where you have to horizontal scroll.

File: asserts/ifacedecls_test.go [hunk 13]
```diff
@@ -266,5 +295,231 @@ foo: [{p: "/foo/x"}, {p: "/foo/y"}]
...
+	tests := []struct {
+		stanza string
+		err    string
+	}{
+		{`iface: foo`, `plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
+		{`iface:
+  - allow`, `plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
+		{`iface:
+  allow-installation: foo`, `allow-installation in plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
+		{`iface:
```
Comment (jdstrand) [inline asserts/ifacedecls_test.go]: Can you add a deny-installation test?

File: asserts/ifacedecls_test.go [hunk 14]
```diff
@@ -266,5 +295,231 @@ foo: [{p: "/foo/x"}, {p: "/foo/y"}]
...
+      - foo`, `slot-snap-type in allow-connection in plug rule for interface "iface" contains an invalid element: "foo"`},
+		{`iface:
+  allow-connection:
+    slot-snap-type:
+      - xapp`, `slot-snap-type in allow-connection in plug rule for interface "iface" contains an invalid element: "xapp"`},
+		{`iface:
+  allow-connection:
+    slot-snap-ids:
+      - foo`, `allow-connection in plug rule for interface "iface" must specify at least one of plug-attributes, slot-attributes, slot-snap-type, slot-publisher-id, slot-snap-id`},
+		{`iface:
```
Comment (jdstrand) [inline asserts/ifacedecls_test.go]: Can you add a deny-connection, allow-auto-connection and deny-auto-connection test? Not saying we have to do all the tests that you did for allow-connection here for each these, but just a test for each to help future-proof would be nice.

File: asserts/snap_asserts_test.go [hunk 15]
```diff
@@ -154,6 +157,44 @@ func (sds *snapDeclSuite) TestDecodeInvalid(c *C) {
...
+publisher-id: dev-id1
+plugs:
+  interface1:
+    allow-auto-connection:
+      slot-attributes:
+        a1: /foo/.*
+  interface2:
+    allow-connection:
+      slot-attributes:
+        a2: /foo/.*
```
Comment (jdstrand) [inline asserts/snap_asserts_test.go]: Can we also add similar deny-connection, deny-auto-connection, allow-installation and deny-installation tests? Also have slot-publisher-id, slot-snap-id, slot-snap-type and plug-attributes represented here would be good (ie, having a valid example of each field in this toplevel snap declaration test is useful).

File: asserts/ifacedecls_test.go [hunk 16]
```diff
@@ -266,5 +295,231 @@ foo: [{p: "/foo/x"}, {p: "/foo/y"}]
...
+}
+
+func (s *plugSlotRulesSuite) TestCompilePlugRuleErrors(c *C) {
+	tests := []struct {
+		stanza string
+		err    string
+	}{
+		{`iface: foo`, `plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
+		{`iface:
+  - allow`, `plug rule for interface "iface" must be a map or one of the shortcuts 'true' or 'false'`},
```
Comment (pedronis) [inline asserts/ifacedecls_test.go]: yes, you are reading this right and yes it's a generalisation I thought considering the base declaration

File: asserts/ifacedecls.go [hunk 17]
```diff
@@ -220,7 +222,263 @@ func compileAttributeConstraints(constraints interface{}) (*AttributeConstraints
...
+
+var (
+	defaultOutcome = map[string]interface{}{
+		"allow-installation":    "true",
+		"allow-connection":      "true",
+		"allow-auto-connection": "true",
+		"deny-installation":     "false",
+		"deny-connection":       "false",
+		"deny-auto-connection":  "false",
+	}
```
Comment (pedronis) [inline asserts/ifacedecls.go]: I don't know myself either, anyway that's why I coded this way very declaritvely/explicitly, we can also have different approaches on the slots vs plugs side if needed

the other thing we were discussing with Gustavo, is how much we want to move to have entries for everything in the base declaration

File: asserts/ifacedecls_test.go (hunk 16 above)
Comment (pedronis) [inline asserts/ifacedecls_test.go]: we should document it in the doc

Comment (jdstrand) [inline asserts/ifacedecls_test.go]: Ok, that makes a lot of sense then. Thanks!

**Comment by matiasb** [comment]
Looking good, changes make sense to me.

File: asserts/ifacedecls.go (hunk 12 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: I'm rewrapping some

File: asserts/snap_asserts_test.go (hunk 15 above)
Comment (pedronis) [inline asserts/snap_asserts_test.go]: allow/deny-installation doesnt make a lot of sense at the snap-declaration level, I fear it would be a bit confusing, I can add the other kind of fields and subrules though

File: asserts/ifacedecls_test.go [hunk 18]
```diff
@@ -68,6 +73,30 @@ func (s *attrConstraintsSuite) TestSimple(c *C) {
...
+
+	err = cstrs.Check(map[string]interface{}{
+		"bar": "BARR",
+	})
+	c.Check(err, ErrorMatches, `attribute "bar" value "BARR" does not match \^\(BAR|BAZ\)\$`)
+
+	err = cstrs.Check(map[string]interface{}{
+		"bar": "BBAZ",
+	})
+	c.Check(err, ErrorMatches, `attribute "bar" value "BAZZ" does not match \^\(BAR|BAZ\)\$`)
```
Comment (pedronis) [inline asserts/ifacedecls_test.go]: Done

File: asserts/ifacedecls_test.go (hunk 14 above)
Comment (pedronis) [inline asserts/ifacedecls_test.go]: Done

File: asserts/snap_asserts_test.go (hunk 15 above)
Comment (jdstrand) [inline asserts/snap_asserts_test.go]: Well, if the base declaration has a blanket deny-installation then we would want an allow-installation in the snap declaration to override that, no? This is how I thought we would handle the hardcoded logic for lxd and docker.

**Comment by jdstrand** [comment]
These latest changes look great, thank you. :)
